File: 203. Remove Linked List Elements.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def removeElements(self, head, val):
        """
        :type head: ListNode
        :type val: int
        :rtype: ListNode
        """
        dummy = ListNode(-1)
        dummy.next = head
        p = head
        pre = dummy
        
        while p:
            if p.val == val:
                pre.next = p.next
                del p
                p = pre.next
            else:
                p = p.next
                pre = pre.next
        return dummy.next

File: 203. Remove Linked List Elements.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        ListNode* dummy = new ListNode(-1);
        dummy->next = head;
        ListNode* p = head;
        ListNode* pre = dummy;
        
        while (p != NULL)
        {
            if (p->val == val)
            {
                pre->next = p->next;
                delete p;
                p = pre->next;
            }
            else
            {
                p = p->next;
                pre = pre->next;
            }
        }
        return dummy->next;
    }
};
